pydantic/schema.py

```diff
@@ -662,6 +691,7 @@ def field_singleton_schema(  # noqa: C901 (ignore complexity)
     model_name_map: Dict[Type['BaseModel'], str],
     schema_overrides: bool = False,
     ref_prefix: Optional[str] = None,
+    known_models: Set[Type['BaseModel']],
```

Comment:
this function is **MASSIVE** is there anything we can do to simplify it?

Reply:
Let me check what I can do...

Reply:
if it's not possible or too difficult, don't worry. Maybe better to do in a separate PR anyway.

Reply:
Hmm, I agree it's big. But I don't really know what to put outside, I don't see much code repetition as it handles several different cases.

Maybe the last section that checks if it's a model, under `if issubclass(field_type, pydantic.BaseModel):` could be in a function...

But it would require several parameters, would have to modify some of them (which is a bit difficult to debug/refactor later) and would require an extra `if` to check if it returned a value (to return it) or not, so I'm not sure.

Do you want me to put it outside? Or later?

Reply:
let's do it (or not do it) later.